internal/docstore/docstore.go

```diff
@@ -0,0 +1,216 @@
+// retrieved document. If fps is present, only the given field paths are
+// set. In both cases, other fields of doc are not touched.
+//
+// A field path is a dot-separated sequence of UTF-8 field names. A field path
+// can select top level fields or elements of sub-documents. There is no way to
+// select a single list element.
+func (l *ActionList) Get(doc Document, fps ...string) *ActionList {
+	return l.add(&Action{
+		kind:       driver.Get,
+		doc:        doc,
+		fieldpaths: fps,
+	})
+}
+
+// Update applies Mods to doc, which must exist.
+// Only the key and revision fields of doc are used.
+//
+// A modification will create a field if it doesn't exist.
+//
+// No field path in mods can be a prefix of another. (It makes no sense
+// to, say, set foo but increment foo.bar.)
+//
+// Update does not modify its doc argument. To obtain the new value of the document,
+// call Get after calling Update.
+func (l *ActionList) Update(doc Document, mods Mods) *ActionList {
+	return l.add(&Action{
+		kind: driver.Update,
+		doc:  doc,
+		mods: mods,
+	})
+}
+
+// Mods is a map from field paths to modifications.
+// A field path is a dot-separated sequence of UTF-8 field names.
+// See ActionList.Update.
+type Mods map[string]interface{}
+
+// Do executes the action list. If all the actions executed successfully, Do returns
+// (number of actions, nil). If any failed, Do returns the number of successful
+// actions and an error. In general there is no way to know which actions succeeded,
+// but the error will contain as much information as possible about the failures.
+func (l *ActionList) Do(ctx context.Context) (int, error) {
+	var das []*driver.Action
+	for _, a := range l.actions {
+		d, err := a.toDriverAction()
+		if err != nil {
+			return 0, wrapError(l.coll.driver, err)
+		}
+		das = append(das, d)
+	}
+	n, err := l.coll.driver.RunActions(ctx, das)
+	return n, wrapError(l.coll.driver, err)
+}
+
+func (a *Action) toDriverAction() (*driver.Action, error) {
+	ddoc, err := driver.NewDocument(a.doc)
+	if err != nil {
+		return nil, err
+	}
+	d := &driver.Action{Kind: a.kind, Doc: ddoc}
+	if a.fieldpaths != nil {
+		d.FieldPaths = make([][]string, len(a.fieldpaths))
+		for i, s := range a.fieldpaths {
+			fp, err := parseFieldPath(s)
+			if err != nil {
+				return nil, err
+			}
+			d.FieldPaths[i] = fp
+		}
+	}
+	if a.mods != nil {
+		// Convert mods from a map to a slice of (fieldPath, value) pairs.
+		// The map is easier for users to write, but the slice is easier
+		// to process.
+		// TODO(jba): check for prefix
+		for k, v := range a.mods {
+			fp, err := parseFieldPath(k)
+			if err != nil {
+				return nil, err
+			}
+			d.Mods = append(d.Mods, driver.Mod{fp, v})
+		}
+	}
+	return d, nil
+}
+
+func parseFieldPath(s string) ([]string, error) {
+	if !utf8.ValidString(s) {
+		return nil, gcerr.Newf(gcerr.InvalidArgument, nil, "invalid UTF-8 field path %q", s)
+	}
+	fp := strings.Split(s, ".")
```

Comment:
What happens if I have a map with a string key with "." in it?

Reply:
Fields cannot have dots. I clarified this in the `Document` docstring.

If the map key consists solely of ".", or has other invalid uses of ".", then parseFieldPath should catch it.